ExecutionEngineException due to race between thread creation and shutdown

From https://github.com/dotnet/runtime/pull/31756 on `Libraries Test Run release mono Linux x64 Debug` CI leg.

Specifically: netcoreapp5.0-Linux-Debug-x64-Mono_release-Debian.9.Amd64.Open

https://helix.dot.net/api/2019-06-17/jobs/b8f6a868-71db-4bc3-9142-3adb3b512af6/workitems/System.Net.Http.Functional.Tests/console
```text
  Discovering: System.Net.Http.Functional.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  System.Net.Http.Functional.Tests (found 0 of 853 test cases)
=== TEST EXECUTION SUMMARY ===
   System.Net.Http.Functional.Tests.dll  Total: 0

Unhandled Exception:
System.ExecutionEngineException: Couldn't create thread. Runtime is shutting down.
   at System.Threading.Thread.Start()
   at System.Threading.PortableThreadPool.WorkerThread.TryCreateWorkerThread()
   at System.Threading.PortableThreadPool.WorkerThread.MaybeAddWorkingWorker()
   at System.Threading.PortableThreadPool.RequestWorker()
   at System.Threading.ThreadPool.RequestWorkerThread()
   at System.Threading.ThreadPoolWorkQueue.EnsureThreadRequested()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
   at System.Threading.Thread.StartCallback()
[ERROR] FATAL UNHANDLED EXCEPTION: System.ExecutionEngineException: Couldn't create thread. Runtime is shutting down.
   at System.Threading.Thread.Start()
   at System.Threading.PortableThreadPool.WorkerThread.TryCreateWorkerThread()
   at System.Threading.PortableThreadPool.WorkerThread.MaybeAddWorkingWorker()
   at System.Threading.PortableThreadPool.RequestWorker()
   at System.Threading.ThreadPool.RequestWorkerThread()
   at System.Threading.ThreadPoolWorkQueue.EnsureThreadRequested()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
   at System.Threading.Thread.StartCallback()
~/work/B1C609A0/w/9DFE0880/e
----- end Thu Feb 6 03:23:51 UTC 2020 ----- exit code 1 ----------------------------------------------------------
Looking around for any Linux dump...
... found no dump in /home/helixbot/work/B1C609A0/w/9DFE0880/e
+ export _commandExitCode=1
```
https://dev.azure.com/dnceng/public/_build/results?buildId=509474&view=ms.vss-test-web.build-test-results-tab&runId=16167872&paneView=debug&resultId=168275

--- a/src/mono/mono/metadata/threads.c
+++ b/src/mono/mono/metadata/threads.c
@@ -1356,8 +1356,12 @@ create_thread (MonoThread *thread, MonoInternalThread *internal, MonoObject *sta
 	mono_threads_lock ();
 	if (shutting_down && !(flags & MONO_THREAD_CREATE_FLAGS_FORCE_CREATE)) {
 		mono_threads_unlock ();
-		mono_error_set_execution_engine (error, "Couldn't create thread. Runtime is shutting down.");
-		return FALSE;
+		/* We're already shutting down, don't create the new
+		 * thread. Instead detach and exit from the current thread.
+		 * Don't expect mono_threads_set_shutting_down to return.
+		 */
+		mono_threads_set_shutting_down ();
+		g_assert_not_reached ();
 	}
 	if (threads_starting_up == NULL) {
 		threads_starting_up = mono_g_hash_table_new_type_internal (NULL, NULL, MONO_HASH_KEY_VALUE_GC, MONO_ROOT_SOURCE_THREADING, NULL, "Thread Starting Table");

